en/lessons/basics/modules.md

```diff
@@ -1,316 +1,195 @@
+are groupings of entities with some or all exported. Components are graphs of
+modules that can be made available in other codebases or executed via an
+entrypoint module.
 
-- `moduledoc` — Documents the current module.
-- `doc` — Documentation for functions and macros.
-- `behaviour` — Use an OTP or user-defined behaviour.
+## Module Syntax
 
-## Structs
+Module syntax is driven by 2 keywords, `module` and `where`. The `module` keyword
+precedes the modules name, which must match the filename, and optionally a
+collection of entities to export. The `where` keyword ends the module
+declaration and indicates the beginning of the module.
 
-Structs are special maps with a defined set of keys and default values.
-A struct must be defined within a module, which it takes its name from.
-It is common for a struct to be the only thing defined within a module.
+Below is an example of a module declaration exposing all entities.
 
-To define a struct we use `defstruct` along with a keyword list of fields and default values:
+```haskell
+-- src/Example.hs
+module Example where
 
-```elixir
-defmodule Example.User do
-  defstruct name: "Sean", roles: []
-end
-```
-
-Let's create some structs:
-
-```elixir
-iex> %Example.User{}
-%Example.User<name: "Sean", roles: [], ...>
-
-iex> %Example.User{name: "Steve"}
-%Example.User<name: "Steve", roles: [], ...>
-
-iex> %Example.User{name: "Steve", roles: [:manager]}
-%Example.User<name: "Steve", roles: [:manager]>
-```
+id :: a -> a
+id a = a
 
-We can update our struct just like we would a map:
-
-```elixir
-iex> steve = %Example.User{name: "Steve"}
-%Example.User<name: "Steve", roles: [...], ...>
-iex> sean = %{steve | name: "Sean"}
-%Example.User<name: "Sean", roles: [...], ...>
+compose :: (b -> c) -> (a -> b) -> (a -> c)
+compose f g = \a -> f (g a)
 ```
 
-Most importantly, you can match structs against maps:
+Here is the same module, but residing at a different filepath.
 
-```elixir
-iex> %{name: "Sean"} = sean
-%Example.User<name: "Sean", roles: [...], ...>
-```
+```haskell
+-- src/Data/Examples/Example.hs
+module Data.Examples.Example where
 
-As of Elixir 1.8 structs include custom introspection.
-To understand what this means and how we are to use it let us inspect our `sean` capture:
+id :: a -> a
+id a = a
 
-```elixir
-iex> inspect(sean)
-"%Example.User<name: \"Sean\", roles: [...], ...>"
+compose :: (b -> c) -> (a -> b) -> (a -> c)
+compose f g = \a -> f (g a)
 ```
 
-All of our fields are present which is okay for this example but what if we had a protected field we didn't want to include?
-The new `@derive` feature let's us accomplish just this!
-Let's update our example so `roles` are no longer included in our output:
+Modules can export a subset of the entities they contain. In the example below
+only the type `Identity` (and its constructor) and `id` are exported, `compose`
+is completely isolated to the local module scope.
```

Comment:
It's unfortunate, but for the flow of the lessons, I believe we're going to restrain from code examples with types and typeclasses here, given the lesson is right at the start of the tutorial. However, I think it would be beneficial to the reader for you to quickly mention them (e.g. in your listing of entities / as a note). 

The plan is to introduce how to import and export each entity (except for declarations) when we reach them.

Reply:
Really? I am not sure I agree with this decision. Perhaps typeclasses (which I believe are not included in any code examples), but I think seeing a type exported and imported will be more helpful to the user than the confusion that might arise. Type declarations are one of the first things one learns about in haskell.

Reply:
Reconsidering it, I'd agree with you that your example has little potential to cause confusion, and comes with some benefits. However, a note (e.g. `__Note__: you'll learn about using and defining types in Haskell in [types](../types) and [algebraic data types](../algebraic-data-types).`) at the end of a codeblock may help the reader even more.

Reply:
Yeah, I agree, that's a good idea.